File: stackunderflow/etl/normalize/opencode.py
"""OpenCode normalizer.

OpenCode persists chats in a SQLite database (``opencode*.db`` under
``$XDG_DATA_HOME/opencode/``). The ``message`` table's ``data`` column
holds the per-turn payload as JSON, with the canonical token block at
``data.tokens``:

    { "role": "assistant",
      "modelID": "...",
      "tokens": {
        "input": ...,
        "output": ...,
        "reasoning": ...,
        "cache": {"read": ..., "write": ...}
      },
      "cost": ... }

Per the codeburn catalog the canonical mapping is:

* ``input  = tokens.input``
* ``output = tokens.output + tokens.reasoning``  (fold reasoning into output)
* ``cache_read   = tokens.cache.read``
* ``cache_create = tokens.cache.write``

Model id is preserved verbatim from ``data.modelID`` (the adapter
should already lift it onto the row column). ``cost`` is not
consumed — we recompute via ``compute_cost`` for parity, preserving
the source-side cost in ``raw_extras``.
"""

from __future__ import annotations

import json
from collections.abc import Iterable

from stackunderflow.infra.costs import RATE_CARD

from .base import (
    COST_SOURCE_RATE_CARD,
    COST_SOURCE_UNKNOWN,
    Normalizer,
)

_DEFAULT_MODEL = "opencode-auto"
_RAW_EXTRAS_FIELDS = ("modelID", "providerID", "embeddedCost")


class OpenCodeNormalizer(Normalizer):
    provider_name = "opencode"

    def normalize(self, msg_row: dict) -> Iterable[dict]:
        role = str(msg_row.get("role") or "")
        if role != "assistant":
            return

        canonical = _canonical_tokens(msg_row)
        if canonical is None:
            return

        if all(canonical[k] == 0 for k in ("input", "output", "cache_read", "cache_create")):
            return

        model = str(msg_row.get("model") or "") or _DEFAULT_MODEL

        cost_source = (
            COST_SOURCE_RATE_CARD
            if model in RATE_CARD
            else COST_SOURCE_UNKNOWN
        )

        raw_extras = _extras_from_raw_json(msg_row.get("raw_json"))

        yield self._build_event(
            msg_row,
            input_tokens=canonical["input"],
            output_tokens=canonical["output"],
            cache_read_tokens=canonical["cache_read"],
            cache_create_tokens=canonical["cache_create"],
            cost_source=cost_source,
            model=model,
            raw_extras=raw_extras,
        )


def _canonical_tokens(msg_row: dict) -> dict[str, int] | None:
    """Return canonical 4-key shape with reasoning folded into output."""
    raw = _raw_tokens(msg_row)
    if raw is not None:
        cache = raw.get("cache") if isinstance(raw.get("cache"), dict) else {}
        return {
            "input": _safe_int(raw.get("input")),
            "output": _safe_int(raw.get("output")) + _safe_int(raw.get("reasoning")),
            "cache_read": _safe_int(cache.get("read") if isinstance(cache, dict) else 0),
            "cache_create": _safe_int(cache.get("write") if isinstance(cache, dict) else 0),
        }
    return {
        "input": int(msg_row.get("input_tokens") or 0),
        "output": int(msg_row.get("output_tokens") or 0),
        "cache_read": int(msg_row.get("cache_read_tokens") or 0),
        "cache_create": int(msg_row.get("cache_create_tokens") or 0),
    }


def _raw_tokens(msg_row: dict) -> dict | None:
    """Locate the OpenCode tokens block on msg_row or raw_json."""
    direct = msg_row.get("tokens")
    if isinstance(direct, dict):
        return direct
    payload = _safe_load_raw(msg_row.get("raw_json"))
    if isinstance(payload, dict):
        # Canonical persistence — the message.data JSON has tokens at
        # the top level.
        tokens = payload.get("tokens")
        if isinstance(tokens, dict):
            return tokens
        # If the raw_json includes a wrapping ``data`` key (some adapter
        # versions persist the row that way), unwrap once.
        data = payload.get("data")
        if isinstance(data, dict):
            tokens = data.get("tokens")
            if isinstance(tokens, dict):
                return tokens
    return None


def _safe_int(value: object) -> int:
    try:
        return max(int(value or 0), 0)
    except (TypeError, ValueError):
        return 0


def _extras_from_raw_json(raw_json: object) -> dict | None:
    payload = _safe_load_raw(raw_json)
    if not isinstance(payload, dict):
        return None
    inner = payload.get("data") if isinstance(payload.get("data"), dict) else payload
    out: dict = {}
    for key in _RAW_EXTRAS_FIELDS:
        val = inner.get(key) if isinstance(inner, dict) else None
        if val is not None and val != "":
            out[key] = val
    cost = inner.get("cost") if isinstance(inner, dict) else None
    if cost is not None and "embeddedCost" not in out:
        out["embeddedCost"] = cost
    return out or None


def _safe_load_raw(raw_json: object) -> object | None:
    if isinstance(raw_json, dict):
        return raw_json
    if not isinstance(raw_json, str | bytes | bytearray):
        return None
    try:
        if isinstance(raw_json, bytes | bytearray):
            return json.loads(raw_json.decode("utf-8", errors="replace"))
        return json.loads(raw_json)
    except (json.JSONDecodeError, ValueError, UnicodeDecodeError):
        return None

File: rust/crates/stax-etl/src/normalize/opencode.rs
//! OpenCode — port of `etl/normalize/opencode.py`.
//!
//! The per-turn payload lives in the `message.data` JSON column with tokens at
//! `data.tokens`: `{input, output, reasoning, cache: {read, write}}`. Reasoning
//! folds into output; the cache block splits into read/write. `cost` is
//! preserved as `embeddedCost` and not consumed.

use stax_core::queries::pyjson::Value as PyValue;

use super::base::{EventSpec, NormalizeContext, Normalizer, UsageEvent};
use super::row::{MsgRow, PyRaise, as_dict, int_or_zero, safe_int, safe_load_raw, str_or_empty};
use super::text::{collect_fields, unwrap_envelope};
use crate::pricing::Tokens;

const DEFAULT_MODEL: &str = "opencode-auto";
const RAW_EXTRAS_FIELDS: [&str; 3] = ["modelID", "providerID", "embeddedCost"];

/// The `opencode` normalizer.
#[derive(Debug, Clone, Copy, Default)]
pub struct OpenCodeNormalizer;

impl Normalizer for OpenCodeNormalizer {
    fn provider_name(&self) -> &'static str {
        "opencode"
    }

    fn normalize(&self, ctx: &NormalizeContext, row: &MsgRow) -> Result<Vec<UsageEvent>, PyRaise> {
        if str_or_empty(row, "role") != "assistant" {
            return Ok(Vec::new());
        }

        let canonical = canonical_tokens(row)?;
        if canonical.input == 0
            && canonical.output == 0
            && canonical.cache_read == 0
            && canonical.cache_creation == 0
        {
            return Ok(Vec::new());
        }

        let model = match str_or_empty(row, "model") {
            empty if empty.is_empty() => DEFAULT_MODEL.to_string(),
            model => model,
        };
        let cost_source = ctx.rate_card_or_unknown(&model);

        Ok(vec![
            self.build_event(
                ctx,
                row,
                EventSpec::new(
                    canonical.input,
                    canonical.output,
                    canonical.cache_read,
                    canonical.cache_creation,
                    cost_source,
                )
                .model(model)
                .raw_extras(extras_from_raw_json(row.get("raw_json"))),
            ),
        ])
    }
}

/// The canonical four with reasoning folded into output.
///
/// **`reasoning_tokens` stays 0 here** even though OpenCode reports a separable
/// count — unlike codex and droid, which surface theirs. That asymmetry is the
/// reference's, not a porting slip; it is recorded rather than corrected.
fn canonical_tokens(row: &MsgRow) -> Result<Tokens, PyRaise> {
    if let Some(raw) = raw_tokens(row) {
        let cache = as_dict(raw.get("cache"));
        return Ok(Tokens {
            input: safe_int(raw.get("input"))?,
            output: safe_int(raw.get("output"))? + safe_int(raw.get("reasoning"))?,
            cache_read: safe_int(cache.and_then(|c| c.get("read")))?,
            cache_creation: safe_int(cache.and_then(|c| c.get("write")))?,
        });
    }
    Ok(Tokens {
        input: int_or_zero(row, "input_tokens")?,
        output: int_or_zero(row, "output_tokens")?,
        cache_read: int_or_zero(row, "cache_read_tokens")?,
        cache_creation: int_or_zero(row, "cache_create_tokens")?,
    })
}

/// The tokens block: on the row, at the payload's top level, or one `data`
/// envelope down.
fn raw_tokens(row: &MsgRow) -> Option<PyValue> {
    if let Some(direct) = as_dict(row.get("tokens")) {
        return Some(direct.clone());
    }
    let payload = safe_load_raw(row.get("raw_json"))?;
    if !matches!(payload, PyValue::Object(_)) {
        return None;
    }
    if let Some(tokens) = as_dict(payload.get("tokens")) {
        return Some(tokens.clone());
    }
    let data = as_dict(payload.get("data"))?;
    as_dict(data.get("tokens")).cloned()
}

fn extras_from_raw_json(raw_json: Option<&PyValue>) -> Option<PyValue> {
    let payload = safe_load_raw(raw_json)?;
    if !matches!(payload, PyValue::Object(_)) {
        return None;
    }
    let inner = unwrap_envelope(&payload, "data");
    let mut out = collect_fields(inner, &RAW_EXTRAS_FIELDS);
    if let Some(cost) = inner.get("cost")
        && !matches!(cost, PyValue::Null)
        && !out.iter().any(|(name, _)| name == "embeddedCost")
    {
        out.push(("embeddedCost".to_string(), cost.clone()));
    }
    (!out.is_empty()).then_some(PyValue::Object(out))
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::normalize::test_support::{assistant_row, ctx};

    fn opencode_row(data: &str) -> MsgRow {
        assistant_row("opencode", "claude-sonnet-4-5-20250929")
            .with("raw_json", PyValue::Str(data.to_string()))
    }

    #[test]
    fn reasoning_folds_into_output_and_the_cache_block_splits() {
        let row = opencode_row(
            r#"{"tokens": {"input": 100, "output": 20, "reasoning": 5,
                "cache": {"read": 7, "write": 9}}, "cost": 0.42,
                "modelID": "claude-sonnet-4-5-20250929"}"#,
        );
        let events = OpenCodeNormalizer.normalize(&ctx(), &row).unwrap();
        assert_eq!(events[0].input_tokens, 100);
        assert_eq!(events[0].output_tokens, 25);
        assert_eq!(events[0].cache_read_tokens, 7);
        assert_eq!(events[0].cache_create_tokens, 9);
        // Reported separably by the provider, deliberately not attributed here.
        assert_eq!(events[0].reasoning_tokens, 0);
        assert_eq!(
            events[0].raw_extras.as_deref(),
            Some(r#"{"modelID": "claude-sonnet-4-5-20250929", "embeddedCost": 0.42}"#)
        );
    }

    #[test]
    fn a_data_envelope_is_unwrapped_once() {
        let row = opencode_row(r#"{"data": {"tokens": {"input": 11}, "providerID": "anthropic"}}"#);
        let events = OpenCodeNormalizer.normalize(&ctx(), &row).unwrap();
        assert_eq!(events[0].input_tokens, 11);
        assert_eq!(
            events[0].raw_extras.as_deref(),
            Some(r#"{"providerID": "anthropic"}"#)
        );
    }

    #[test]
    fn a_missing_cache_block_is_zero_not_a_raise() {
        let row = opencode_row(r#"{"tokens": {"input": 3}}"#);
        let events = OpenCodeNormalizer.normalize(&ctx(), &row).unwrap();
        assert_eq!(events[0].cache_read_tokens, 0);
        assert_eq!(events[0].cache_create_tokens, 0);
    }
}
